Proc/Spawn refactor

rgrinberg opened the pull request:
First move everything to Spawn. This allows us tho share code

Then fix `exec` on Windows. On windows, `exec` wouldn't terminate. This doesn't
match the behavior of `exec` on Unix.

aalekseyev commented:
The change looks good to me. I pushed a commit to move `Unix.sigprocmask` back before redirections. 
I don't know if that's important in practice, but if I imagine we want that call to be as close as possible to `fork` so that we can't have a parent's signal handler running in the child process.

I don't understand the CI failure. That doesn't seem related from the first glance.

rgrinberg commented:
This is somehow broken on Windows now. Will investigate.

```
(cd _build/.sandbox/4b604ba5e76997a2c8a3f4d748db8755/default && C:\cygwin\bin\git.exe --no-pager diff --no-index --color=always -u --ignore-cr-at-eol test/blackbox-tests/test-cases/windows-diff.t/run.t test/blackbox-tests/test-cases/windows-diff.t/run.t.corrected)
diff --git a/test/blackbox-tests/test-cases/windows-diff.t/run.t b/test/blackbox-tests/test-cases/windows-diff.t/run.t.corrected
index d73f62f..6f81c29 100755
--- a/test/blackbox-tests/test-cases/windows-diff.t/run.t
+++ b/test/blackbox-tests/test-cases/windows-diff.t/run.t.corrected
@@ -1,18 +1,23 @@
   $ echo 'Hello, world!' > hello.expected
   $ dune runtest
+  Entering directory 'D:\a\dune\dune'
+  Error: This path is internal to dune:
+  _build/.sandbox/4b604ba5e76997a2c8a3f4d748db8755/default/test/blackbox-tests/test-cases/windows-diff.t
+  [1]
```
